File: app/src/main/java/com/settlex/android/ui/common/components/OtpVerificationActivity.java
package com.settlex.android.ui.common.components;

import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.MotionEvent;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.settlex.android.R;
import com.settlex.android.databinding.ActivityOtpVerificationBinding;
import com.settlex.android.ui.auth.activity.SetNewPasswordActivity;
import com.settlex.android.ui.auth.viewmodel.AuthViewModel;
import com.settlex.android.ui.common.util.ProgressLoaderController;
import com.settlex.android.utils.event.Result;
import com.settlex.android.utils.network.NetworkMonitor;
import com.settlex.android.utils.string.StringUtil;
import com.settlex.android.utils.ui.StatusBarUtil;
import com.settlex.android.utils.ui.UiUtil;

import java.util.Objects;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class OtpVerificationActivity extends AppCompatActivity {

    private boolean isConnected = false;

    // dependencies
    private String userEmail;
    private AuthViewModel authViewModel;
    private ActivityOtpVerificationBinding binding;
    private ProgressLoaderController progressLoader;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityOtpVerificationBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        userEmail = getIntent().getStringExtra("email");
        authViewModel = new ViewModelProvider(this).get(AuthViewModel.class);
        progressLoader = new ProgressLoaderController(this);

        setupUiActions();
        observeNetworkStatus();
        observeSendPasswordResetStatus();
        observeVerifyPasswordResetStatus();
    }

    private void setupUiActions() {
        StatusBarUtil.setStatusBarColor(this, R.color.white);
        handleIntent();
        setupOtpInputWatcher();
        startResendOtpCooldown();
        maskAndDisplayUserEmail();

        binding.btnBackBefore.setOnClickListener(v -> finish());
        binding.btnResendOtp.setOnClickListener(v -> resendOtpCode());
        binding.btnConfirm.setOnClickListener(v -> verifyPasswordReset());
        binding.btnHelp.setOnClickListener(v -> StringUtil.showNotImplementedToast(this));
    }

    private void handleIntent() {
        String intent = getIntent().getStringExtra("session");
        if (intent != null) {
            binding.btnHelp.setVisibility(View.GONE);
        }
    }

    private void observeNetworkStatus() {
        NetworkMonitor.getNetworkStatus().observe(this, isConnected -> {
            if (!isConnected) showNoInternetDialog();
            this.isConnected = isConnected;
        });
    }

    private void showNoInternetDialog() {
        UiUtil.showNoInternetAlertDialog(this);
    }

    private void observeVerifyPasswordResetStatus() {
        authViewModel.getVerifyPasswordResetLiveData().observe(this, event -> {
            Result<String> result = event.getContentIfNotHandled();
            if (result != null) {
                switch (result.getStatus()) {
                    case LOADING -> progressLoader.show();
                    case SUCCESS -> onVerifyPasswordResetStatusSuccess();
                    case FAILURE -> onVerifyPasswordResetStatusError(result.getErrorMessage());
                }
            }
        });
    }

    private void onVerifyPasswordResetStatusSuccess() {
        Intent intent = new Intent(this, SetNewPasswordActivity.class);
        intent.putExtra("email", userEmail);
        intent.putExtra("session", getIntent().getStringExtra("session"));
        startActivity(intent);
        finish();

        progressLoader.hide();
    }

    private void onVerifyPasswordResetStatusError(String message) {
        binding.txtOtpFeedback.setText(message);
        binding.txtOtpFeedback.setVisibility(View.VISIBLE);
        progressLoader.hide();
    }

    private void observeSendPasswordResetStatus() {
        authViewModel.getSendPasswordResetCodeLiveData().observe(this, event -> {
            Result<String> result = event.getContentIfNotHandled();
            if (result != null) {
                switch (result.getStatus()) {
                    case LOADING -> progressLoader.show();
                    case SUCCESS -> onSendPasswordResetCodeStatusSuccess();
                    case FAILURE -> onSendPasswordResetCodeStatusError(result.getErrorMessage());
                }
            }
        });
    }

    private void onSendPasswordResetCodeStatusSuccess() {
        startResendOtpCooldown();
        progressLoader.hide();
    }

    private void onSendPasswordResetCodeStatusError(String error) {
        binding.txtOtpFeedback.setText(error);
        binding.txtOtpFeedback.setVisibility(View.VISIBLE);
        progressLoader.hide();
    }

    private void verifyPasswordReset() {
        if (!isConnected) {
            showNoInternetDialog();
            return;
        }
        authViewModel.verifyPasswordReset(userEmail, getEnteredOtpCode());
    }

    private void resendOtpCode() {
        if (!isConnected) {
            showNoInternetDialog();
            return;
        }
        authViewModel.sendPasswordResetCode(userEmail);
    }

    private void setupOtpInputWatcher() {
        binding.otpBox.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void afterTextChanged(Editable s) {
            }

            @Override
            public void onTextChanged(CharSequence otp, int start, int before, int count) {
                if (otp.toString().isEmpty()) binding.txtOtpFeedback.setVisibility(View.GONE);
                updateButtonConfirmState();
            }
        });
    }

    private void updateButtonConfirmState() {
        binding.btnConfirm.setEnabled(isOtpBoxFilled());
    }

    private boolean isOtpBoxFilled() {
        return binding.otpBox.length() == binding.otpBox.getItemCount();
    }

    private String getEnteredOtpCode() {
        return Objects.requireNonNull(binding.otpBox.getText()).toString();
    }

    private void startResendOtpCooldown() {
        binding.btnResendOtp.setEnabled(false);
        binding.btnResendOtp.setTag(binding.btnResendOtp.getText());

        new CountDownTimer(60000, 1000) {
            public void onTick(long millisUntilFinished) {

                int seconds = (int) (millisUntilFinished / 1000);
                String COUNT_DOWN = "Resend in " + seconds;

                if (seconds > 0) {
                    binding.btnResendOtp.setText(COUNT_DOWN);
                }
            }

            public void onFinish() {
                binding.btnResendOtp.setText((CharSequence) binding.btnResendOtp.getTag());
                binding.btnResendOtp.setEnabled(true);
            }
        }.start();
    }

    private void maskAndDisplayUserEmail() {
        String maskedEmail = StringUtil.maskEmail(userEmail);
        binding.txtUserEmail.setText(maskedEmail);
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            View v = getCurrentFocus();
            if (v instanceof EditText) {
                Rect outRect = new Rect();
                v.getGlobalVisibleRect(outRect);
                if (!outRect.contains((int) event.getRawX(), (int) event.getRawY())) {
                    v.clearFocus();
                    hideKeyboard(v);
                }
            }
        }
        return super.dispatchTouchEvent(event);
    }

    private void hideKeyboard(View view) {
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }
}

File: app/src/main/java/com/settlex/android/presentation/auth/forgot_password/OtpVerificationActivity.kt
package com.settlex.android.presentation.auth.forgot_password

import android.content.Intent
import android.os.Bundle
import android.os.CountDownTimer
import android.view.MotionEvent
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.widget.doOnTextChanged
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.settlex.android.R
import com.settlex.android.data.enums.OtpType
import com.settlex.android.data.exception.AppException
import com.settlex.android.databinding.ActivityOtpVerificationBinding
import com.settlex.android.presentation.auth.AuthViewModel
import com.settlex.android.presentation.auth.util.PasswordFlow
import com.settlex.android.presentation.auth.util.PasswordFlowParser
import com.settlex.android.presentation.common.extensions.gone
import com.settlex.android.presentation.common.extensions.maskEmail
import com.settlex.android.presentation.common.extensions.setTextColorRes
import com.settlex.android.presentation.common.extensions.show
import com.settlex.android.presentation.common.state.UiState
import com.settlex.android.presentation.common.util.FocusManager
import com.settlex.android.presentation.common.util.SpannableTextFormatter
import com.settlex.android.util.ui.ProgressDialogManager
import com.settlex.android.util.ui.StatusBar
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class OtpVerificationActivity : AppCompatActivity() {

    private lateinit var binding: ActivityOtpVerificationBinding
    private val authViewModel: AuthViewModel by viewModels()
    private val progressLoader by lazy { ProgressDialogManager(this) }
    private val focusManager by lazy { FocusManager(this) }

    private lateinit var passwordFlow: PasswordFlow
    private lateinit var userEmail: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityOtpVerificationBinding.inflate(layoutInflater)
        setContentView(binding.root)

        passwordFlow = PasswordFlowParser.fromIntent(intent)
        userEmail = intent.getStringExtra("email")!!

        initViews()
        initObservers()
    }

    private fun initObservers() {
        observePasswordResetVerification()
        observeOtpRequestEvent()
    }

    private fun initViews() {
        StatusBar.setColor(this, R.color.colorSurface)
        setupListeners()
        setupInputWatcher()
        startOtpResendCooldownTimer()

        userEmail.maskEmail().also { maskedEmail ->
            binding.tvInfo.text = SpannableTextFormatter(
                this,
                "We sent a verification code to your email $maskedEmail}",
                maskedEmail,
                R.color.colorPrimary
            )
        }

        binding.tvWarning.text = SpannableTextFormatter(
            this,
            "This code expires in 10mins. Never share this code with anyone.",
            "Never share this code with anyone.",
            R.color.colorOnWarningContainer
        )
    }

    private fun setupListeners() = with(binding) {
        toolbar.setNavigationOnClickListener { finish() }
        tvResendOtp.setOnClickListener { resendVerificationCode() }
        btnConfirm.setOnClickListener { onConfirmButtonClicked() }
    }

    // Observers
    private fun observePasswordResetVerification() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                authViewModel.verifyPasswordResetEvent.collect {
                    when (it) {
                        is UiState.Loading -> progressLoader.show()
                        is UiState.Success -> proceedToCreatePassword()
                        is UiState.Failure -> showPasswordResetError(it.exception)
                    }
                }
            }
        }
    }

    private fun proceedToCreatePassword() {
        startActivity(
            Intent(this, CreatePasswordActivity::class.java)
                .putExtra("email", userEmail)
                .putExtra("password_flow", intent.getStringExtra("password_flow"))
        )
        progressLoader.hide()
        finish()
    }

    private fun showPasswordResetError(error: AppException) {
        with(binding) {
            tvOtpFeedback.text = error.message
            tvOtpFeedback.show()

            progressLoader.hide()
        }
    }

    private fun observeOtpRequestEvent() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                authViewModel.otpEvent.collect { state ->
                    when (state) {
                        is UiState.Loading -> progressLoader.show()
                        is UiState.Success -> onPasswordResetCodeSent()
                        is UiState.Failure -> showPasswordResetCodeError(state.exception)
                    }
                }
            }
        }
    }

    private fun onPasswordResetCodeSent() {
        startOtpResendCooldownTimer()
        progressLoader.hide()
    }

    private fun showPasswordResetCodeError(error: AppException) {
        with(binding) {
            tvOtpFeedback.text = error.message
            tvOtpFeedback.show()

            progressLoader.hide()
        }
    }

    private fun onConfirmButtonClicked() {
        when (passwordFlow) {
            is PasswordFlow.ForgotPassword -> verifyPasswordReset()

            else -> Unit
        }
    }

    private fun verifyPasswordReset() {
        authViewModel.verifyPasswordReset(userEmail, getEnteredOtpCode())
    }

    private fun resendVerificationCode() {
        authViewModel.sendVerificationCode(userEmail, OtpType.PASSWORD_RESET)
    }

    private fun setupInputWatcher() = with(binding) {
        otpView.doOnTextChanged { _, _, _, _ ->
            if (!isOtpComplete()) tvOtpFeedback.gone()
            updateConfirmButtonEnabled()
        }
    }

    private fun updateConfirmButtonEnabled() {
        binding.btnConfirm.isEnabled = isOtpComplete()
    }

    private fun isOtpComplete(): Boolean {
        return binding.otpView.length() == binding.otpView.itemCount
    }

    private fun getEnteredOtpCode(): String {
        return binding.otpView.getText().toString()
    }

    private fun startOtpResendCooldownTimer() = with(binding) {
        tvResendOtp.apply {
            isEnabled = false
            setTextColorRes(R.color.colorOnSurfaceVariant)
            tag = tvResendOtp.text
        }

        object : CountDownTimer(60000, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                val seconds = (millisUntilFinished / 1000).toInt()
                if (seconds > 0) "Resend in $seconds sec(s)".also { tvResendOtp.text = it }
            }

            override fun onFinish() {
                tvResendOtp.apply {
                    isEnabled = true
                    setTextColorRes(R.color.colorPrimary)
                    text = tvResendOtp.tag as CharSequence
                }
            }
        }.start()
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        if (focusManager.handleOutsideTouch(event)) return true
        return super.dispatchTouchEvent(event)
    }
}
